---
title: "Supplementary figure 5 MAPK"
author: "David Bradley"
date: "10/1/2020"
output: html_document
---

# GGplot font

```{r}

# https://stackoverflow.com/questions/27689222/changing-fonts-for-graphs-in-r

library(extrafont)

font_import()
loadfonts()       #Register fonts for Windows bitmap output
fonts() 

```

# Stoichiometry in humans of MAPK-targeted sites

```{r}

library(gdata)
library(ggseqlogo)
library(ggplot2)
library(stringr)

Ishi <- read.xls('Ishihama_stoichiometry.xlsx',sheet=2, stringsAsFactors=FALSE)

## Exclude sites with 0 stoichiometry
Ishi <- Ishi[!is.nan(Ishi[,8]),]
Ishi <- Ishi[!Ishi[,8] == 0,]
Ishi_MAPK <- Ishi[grep('MAPK',Ishi[,18]),]

# Doubly phosphorylated peptides are potential confounders and should be removed
Ishi_MAPK <- Ishi_MAPK[!Ishi_MAPK[,3] == '2 Phospho (ST)',]

```

# Obtain abundance values for the MAPK substrates (from PaxDB)

```{r}

# Read in the PaxDB table

human_paxdb <- read.table('Human_stoichiometry_PaxDB_integrated.tsv',sep='\t',stringsAsFactors = FALSE)
human_ensembl <- gsub('9606.','',human_paxdb[,2])

# Read in the table to map UniProt IDs to Ensembl IDs

uniprot_map <- read.table('full_uniprot_2_paxdb.04.2015.tsv', stringsAsFactors = FALSE)
uniprot_map <- uniprot_map[uniprot_map[,1] == '9606',]
human_ensembl <- gsub('9606.','',human_paxdb[,2])
ensembl_uniprot <- uniprot_map[uniprot_map[,3] %in% human_ensembl,c(2,3)]
ensembl_uniprot[,1] <- rapply(strsplit(ensembl_uniprot[,1],split='\\|'), function(x) x[1])
rownames(ensembl_uniprot) <- NULL

# Retrieve Ensembl IDs of MAPK substrates

Ishi_MAPK_ensembl <- Ishi_MAPK[Ishi_MAPK[,7] %in% ensembl_uniprot[,1],]
Ishi_MAPK_ensembl <- Ishi_MAPK_ensembl[,c(7,11,8,9)]
colnames(Ishi_MAPK_ensembl) <- c('Accession','Position','Stoichiometry','Peptide')

Ishi_MAPK_ensembl_ext <- Ishi_MAPK_ensembl
ensembl_vec <- NULL
s = 0

for (i in 1:nrow(Ishi_MAPK_ensembl)) {
  
  i2 <- i+s
  
  accession <- Ishi_MAPK_ensembl[i,1]
  ensembl <- ensembl_uniprot[ensembl_uniprot[,1] %in% accession,2]
  
  # Account for the fact that some accessions map to more than one ensembl ID
  
  if (length(ensembl) != 1) {Ishi_MAPK_ensembl_ext <- rbind(Ishi_MAPK_ensembl_ext[1:i2,],Ishi_MAPK_ensembl_ext[i2,],Ishi_MAPK_ensembl_ext[(i2+1):nrow(Ishi_MAPK_ensembl_ext),])}
  if (length(ensembl) != 1) {s = s+1}
  
  ensembl_vec <- c(ensembl_vec, ensembl)
  
}

Ishi_MAPK_ensembl_ext <- cbind(Ishi_MAPK_ensembl_ext, ensembl_vec)
colnames(Ishi_MAPK_ensembl_ext) <- c('Accession', 'Position', 'Stoichiometry','Peptide','Ensembl')

# Retrieve abundance data

abundance_vec <- NULL

for (i in 1:nrow(Ishi_MAPK_ensembl_ext)) {
  
  ensembl_id <- Ishi_MAPK_ensembl_ext[i,5]
  abundance <- human_paxdb[human_ensembl %in% ensembl_id,3]
  abundance_vec <- c(abundance_vec, abundance)
  
}

# Take the log10 of the abundance values

log_abundance_vec <- log(abundance_vec,base=10)
Ishi_abundance <- cbind(Ishi_MAPK_ensembl_ext, log_abundance_vec)
colnames(Ishi_abundance) <- c('Accession','Position','Stoichiometry','Peptide','Ensembl','Abundance')

```

# Phosphoacceptor

```{r}

# Check whether the presence of an S or T has an impact on the stoichiometry measured 

mapk_fasta <- read.fasta('MAPK_substrates.fasta',seqtype = 'AA')
mapk_fasta_names <- rapply(strsplit(names(mapk_fasta),split='\\|'), function(x) x[2])

error_index <- NULL
disorder_res_vec <- NULL

for (i in 1:nrow(Ishi_abundance)) {
  
  accession <- Ishi_abundance[i,1]
  pos <- Ishi_abundance[i,2]
  
  if (accession=='Q99873') {
    
    accession <- 'Q99873'
    pos <- 307
    
  }
  
  if (accession=='P49750') {
    
    accession <- 'P49750'
    pos <- 829
    
  }
  
  if (accession=='Q03252') {
    
    accession <- 'Q03252'
    pos <- 37
    
  }
  
  target_seq <- mapk_fasta[mapk_fasta_names %in% accession]
  target_seq <- unlist(getSequence(target_seq))
  disorder_res <- target_seq[pos]
  
  # Raise the alarm if the position in our data frame does not map to a phospho-acceptor
  
  if (!disorder_res %in% c('S','T','Y') | length(disorder_res) == 0) {
    stop()
  }
  
  disorder_res_vec <- c(disorder_res_vec, disorder_res)
  
}

```

# Disorder content prediction

```{r}

# We will predict protein disorder by using an AF2-based method where the smoothened surface
# accessibility within a sequence window is used as a proxy for protein disorder.
# The approach is described here, and it is where we source the accessibility data:
# https://github.com/normandavey/ProcessedAlphafold

# For proteins without protein structures we use the SPOT-Disorder-Single predictor

Ishi_disorder_MAPK <- readRDS('Ishi_disorder_MAPK.rds')

#AlphaFold disorder prediction on the basis of smoothened RSA

AF2_accessibility <- readLines('9606.accessibility_windowed_scores.15.tdt')

AF2_vec <- NULL

count = 0

for (i in 1:nrow(Ishi_abundance)) {
  
  print(i)
  
  accession <- Ishi_abundance[i,1]
  pos <- Ishi_abundance[i,2]
  
  if (accession=='Q99873') {
    
    accession <- 'Q99873'
    pos <- 307
    
  }
  
  if (accession=='P49750') {
    
    accession <- 'P49750'
    pos <- 829
    
  }
  
  if (accession=='Q03252') {
    
    accession <- 'Q03252'
    pos <- 37
    
  }
  
  grep_dex <- grep(accession,AF2_accessibility)
  
  # Use SPOT-disorder-single if there is no AF2 structure
  
  if (length(grep_dex) == 0) {disorder_pred <- Ishi_disorder_MAPK[i,8]; AF2_vec <- c(AF2_vec,disorder_pred); 
  count = count+1; next}
  
  disorder_pred <- AF2_accessibility[grep_dex]
  disorder_pred <- rapply(strsplit(disorder_pred,split='\t'), function(x) x[2])
  disorder_pred <- unlist(strsplit(disorder_pred,split=','))
  disorder_pred <- disorder_pred[pos]
  
  if (is.na(disorder_pred)) {disorder_pred <- Ishi_disorder_MAPK[i,8]; AF2_vec <- c(AF2_vec,disorder_pred); 
  count = count+1; next}
  
  AF2_vec <- c(AF2_vec,disorder_pred)
  
}

AF2_vec <- as.numeric(AF2_vec)

#####

# Use 0.55 disorder threshold for binary classification of protein disorder
# https://github.com/normandavey/ProcessedAlphafold

disorder_class_vec <- rep('D',length(AF2_vec))
disorder_class_vec[AF2_vec <= 0.55] <- 'O'

Ishi_disorder <- cbind(Ishi_abundance,disorder_res_vec,AF2_vec,disorder_class_vec)
colnames(Ishi_disorder) <- c(colnames(Ishi_abundance),'Phosphoacceptor','Disorder_score','Disorder_class')

```

# Optimal or minimal

```{r}

# Check whether 'minimal'/'optimal' status affects phosphorylation stoichiometry

# Minimal: S/T-P
# Optimal: P-x-S/T-P

MAPK_sites <- Ishi_disorder[,4]
opt_dex <- substr(MAPK_sites,8,8) == 'P'  & substr(MAPK_sites,5,5) %in% c('P')
min_dex <- substr(MAPK_sites,8,8) == 'P'  & !substr(MAPK_sites,5,5) %in% c('P')

Ishi_disorder_motif <- data.frame(Ishi_disorder,rep(0,nrow(Ishi_disorder)))
colnames(Ishi_disorder_motif) <- c(colnames(Ishi_disorder),'Motif')

Ishi_disorder_motif[opt_dex,10] <- 'Optimal'
Ishi_disorder_motif[min_dex,10] <- 'Minimal'

```

# Check for direct interaction between MAPK and substrate

```{r}

Ishi_disorder_motif_beta <- Ishi_disorder_motif

# Read in the iRefindex data for human:

# N.B. This PPI file is too big (2.9 GB) for me to include on GitHub. Please download the PPI file
# for Homo Saoiens from the iRefIndex website: https://irefindex.vib.be/wiki/index.php/iRefIndex

irefindex <- read.table('9606.mitab.27062020.txt', sep='\t', fill=TRUE)

# Find interactors to MAPK1 (the kinase that was actually used for the stoichiometry measurements)

mapk1_int <- unique(c(irefindex[grep('P28482',irefindex[,1]),2],irefindex[grep('P28482',irefindex[,2]),1]))

MAPK_int <- unique(c(mapk1_int))

# Iterate through each of the substrates and determine the number of shared interactions with MAPK
# Check for direct interaction with one of the subunits

shared_vec <- NULL

for (i in 1:nrow(Ishi_disorder_motif_beta)) {
  
  print(i)
  
  uniprot <- Ishi_disorder_motif_beta[i,1]
  
  uniprot_int <- unique(c(irefindex[grep(uniprot,irefindex[,1]),2],irefindex[grep(uniprot,irefindex[,2]),1]))
  
  shared_int <- intersect(uniprot_int, c('uniprotkb:P28482'))
  
  shared_int_len <- length(shared_int)
  
  shared_int_len[shared_int_len > 0] <- 1
  
  shared_vec <- c(shared_vec, shared_int_len)
  
}

Ishi_disorder_motif_beta_int <- cbind(Ishi_disorder_motif_beta,shared_vec)
colnames(Ishi_disorder_motif_beta_int) <- c(colnames(Ishi_disorder_motif_beta),'Int_num')

```

# Subcellular location of MAPK substrates

```{r}

# Subcellular locations for each MAPK substrate were scraped from UniProt

MAPK_uniprot_scrape <- readLines('MAPK_UniProt_raw.txt')

# Parse the raw UniProt entries to the retrieve the subcellular localisation information

MAPK_target_SC_vec <- NULL

for (i in 1:nrow(Ishi_disorder_motif_beta_int)) {
     
    accession <- Ishi_disorder_motif_beta_int[i,1]
    
    MAPK_targets_start_AC <- grep('AC  ',MAPK_uniprot_scrape)
    MAPK_targets_start_accession <- grep(accession,MAPK_uniprot_scrape)
    
    start <- intersect(MAPK_targets_start_AC,MAPK_targets_start_accession)
    if (length(start) >1) {stop('Start problem')}
    
    end_grep <- grep('FT   ',MAPK_uniprot_scrape)
    end_grep <- end_grep[which((end_grep-start) > 0)]
    end <- end_grep[which((end_grep-start) == min(end_grep-start))]
    
    #####
    
    MAPK_target_scrape <- MAPK_uniprot_scrape[start:end]
    
    MAPK_target_GO <- grep('GO;',MAPK_target_scrape)
    MAPK_target_C <- grep('C:',MAPK_target_scrape)  
    
    MAPK_target_GO_C <- intersect(MAPK_target_GO,MAPK_target_C)
    
    MAPK_target_GO_C_text <- MAPK_target_scrape[MAPK_target_GO_C]
    
    if (length(MAPK_target_GO_C_text) == 0) {location <- 'Unknown'; MAPK_target_SC_vec <- c(MAPK_target_SC_vec,location); next}
    
    GO_compartments <- rapply(strsplit(rapply(strsplit(MAPK_target_GO_C_text,split=' C:'), function(x) x[2]), split=';'), function(x) x[1])
    
    #####
    
    # Assign the location to nucleus (N), cytoplasm (C), or mitochondria (M) depending upon matches to the corresponding GO terms
    
    N <- 0
    C <- 0
    M <- 0
    
    nuclear_terms <- c('nucleus','nucleoplasm','nuclear envelope', 'nuclear speck','ribonucleoprotein complex','nucleolus', 'chromosome')
    cyto_terms <- c('cytoplasm','cytosol')
    mito_terms <- c('mitochondria', 'mitochondrion', 'mitochondrial inner membrane', 'mitochondrial outer membrane', 'mitochondrial membrane')
    
    if (length(intersect(GO_compartments,nuclear_terms)) > 0) {N <- 1}
    if (length(intersect(GO_compartments,cyto_terms)) > 0) {C <- 1}
    if (length(intersect(GO_compartments,mito_terms)) > 0) {M <- 1}
    
    if (N == 1 & C == 1 & M == 1) {location <- 'NCM'}
    if (N == 1 & C == 0 & M == 1) {location <- 'NM'}
    if (N == 1 & C == 1 & M == 0) {location <- 'NC'}
    if (N == 0 & C == 1 & M == 1) {location <- 'CM'}
    if (N == 1 & C == 0 & M == 0) {location <- 'N'}
    if (N == 0 & C == 1 & M == 0) {location <- 'C'}
    if (N == 0 & C == 0 & M == 1) {location <- 'M'}
    if (N == 0 & C == 0 & M == 0) {location <- 'None'}
    
    MAPK_target_SC_vec <- c(MAPK_target_SC_vec,location)
    
}

# Group together infrequent labels to maximise power

MAPK_target_SC_vec <- gsub('^CM$','Other',MAPK_target_SC_vec)
MAPK_target_SC_vec <- gsub('^NCM$','Other',MAPK_target_SC_vec)
MAPK_target_SC_vec <- gsub('^NM$','Other',MAPK_target_SC_vec)
MAPK_target_SC_vec <- gsub('^M$','Other',MAPK_target_SC_vec)
MAPK_target_SC_vec <- gsub('^None$','Other',MAPK_target_SC_vec)
MAPK_target_SC_vec <- gsub('^Unknown$','Other',MAPK_target_SC_vec)

# Add to data frame

MAPK_col <- colnames(Ishi_disorder_motif_beta_int) 
Ishi_disorder_motif_beta_int <- cbind(Ishi_disorder_motif_beta_int, MAPK_target_SC_vec)
colnames(Ishi_disorder_motif_beta_int ) <- c(MAPK_col, 'Location')

```

# Stoichiometry GLM (only try and model the significant relationships that were found for CK2)

```{r}

# Save RDS of data frame

saveRDS(Ishi_disorder_motif_beta_int,'Ishi_disorder_MAPK_github.rds')

# Stoichiometry needs to be between 0 and 1 for GLM to work

Ishi_disorder_motif_beta_int[,3] <- Ishi_disorder_motif_beta_int[,3]/100

# Significance of motif

model_motif <- glm(Stoichiometry ~ Motif, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))
model_null <- glm(Stoichiometry ~ 1, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

motif_F_p_MAPK <- anova(model_null,model_motif, test='F')$'Pr(>F)'[2]

# Significance of abundance

model_abundance <- glm(Stoichiometry ~ Abundance, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))
model_null <- glm(Stoichiometry ~ 1, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

abundance_F_p_MAPK <- anova(model_null,model_abundance, test='F')$'Pr(>F)'[2]

# Significance of disorder score

model_disorder_score <- glm(Stoichiometry ~ Disorder_score, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))
model_null <- glm(Stoichiometry ~ 1, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

disorder_score_F_p_MAPK <- anova(model_null,model_disorder_score, test='F')$'Pr(>F)'[2]

# Significance of phosphoacceptor

model_phosphoacceptor <- glm(Stoichiometry ~ Phosphoacceptor, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))
model_null <- glm(Stoichiometry ~ 1, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

phosphoacceptor_class_F_p_MAPK <- anova(model_null,model_phosphoacceptor, test='F')$'Pr(>F)'[2]

# Significance of subcellular location

model_location <- glm(Stoichiometry ~ Location, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))
model_null <- glm(Stoichiometry ~ 1, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

location_F_p_MAPK <- anova(model_null,model_location, test='F')$'Pr(>F)'[2]

# Significance of physical interaction

model_physical <- glm(Stoichiometry ~ Int_num, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))
model_null <- glm(Stoichiometry ~ 1, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

physical_F_p_MAPK <- anova(model_null,model_physical, test='F')$'Pr(>F)'[2]

# Motif:abundance interaction

model_charge_abundance <- glm(Stoichiometry ~ Motif+Abundance, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

model_charge_abundance_int <- glm(Stoichiometry ~ Motif+Abundance+Motif:Abundance, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

charge_abundance_int_F_p_MAPK <- anova(model_charge_abundance,model_charge_abundance_int, test='F')$'Pr(>F)'[2]

# Disorder:phosphoacceptor interaction

model_charge_phosphoacceptor <- glm(Stoichiometry ~ Disorder_score+Phosphoacceptor, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

model_charge_phosphoacceptor_int <- glm(Stoichiometry ~ Disorder_score+Phosphoacceptor+Disorder_score:Phosphoacceptor, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

disorder_phosphoacceptor_p_MAPK <- anova(model_charge_phosphoacceptor,model_charge_phosphoacceptor_int, test='F')$'Pr(>F)'[2]

## Motif:Physical interaction

model_charge_physical <- glm(Stoichiometry ~ Motif+Int_num, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

model_charge_physical_int <- glm(Stoichiometry ~ Motif+Int_num+Motif:Int_num, data = Ishi_disorder_motif_beta_int, family=quasibinomial(cloglog))

charge_physical_int_F_p_MAPK <- anova(model_charge_physical,model_charge_physical_int, test='F')$'Pr(>F)'[2]

```




