---
title: "Supplementary figure 5 CK2"
author: "David Bradley"
date: "10/1/2020"
output: html_document
---

# GGplot font

```{r}

# https://stackoverflow.com/questions/27689222/changing-fonts-for-graphs-in-r

library(extrafont)

font_import()
loadfonts()       #Register fonts for Windows bitmap output
fonts() 

```

# Set up the scoring function for the CK2 substrate quality

```{r}

library(readr)
library(seqinr)
library(ggseqlogo)
library(gdata)

# Data comes from Bachmann et al., 2019 (https://www.biorxiv.org/content/10.1101/822668v3)

prot_mapper <- read_csv('export.csv')

# Kinases only

prot_mapper <- prot_mapper[unlist(prot_mapper[,5]) == TRUE,]

# Retrieve upstream kinase, substrate, and substrate position

ksr <- prot_mapper[,c(3,4,6,9,10)]

# We exclude KSRs that are supported only by the text-mining tools

curated <- c('psp','signor','pid','bel','reactome')

source_list <- strsplit(unlist(ksr[,5]),split=',')

number_curated <- lapply(source_list, function(x) length(intersect(x,curated)))

ksr <- ksr[which(number_curated > 0),]

# Correct formatting problems

for (i in 1:nrow(ksr)) {
 
  if(is.na(ksr[i,2])) {ksr[i,2] <- ksr[i,1]} 
  
}

# Retrieve CSNK2A1 and CSNK2A2 substrates

csnk2a1_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A1') == TRUE),]
csnk2a2_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A2') == TRUE),]
csnk2a_ksr <- unique(rbind(csnk2a1_ksr,csnk2a2_ksr)) 

hs_rp <- read.fasta('hs_proteome_uniprot.fasta', seqtype='AA')
hs_ids <- rapply(strsplit(names(hs_rp),split='\\|'),function(x) x[2])

# Retrieve the sequences

psite_vec <- NULL
index_vec <- NULL

for (i in 1:nrow(csnk2a_ksr)) {
  
  print(i)
  
  # Retrieve the first element only
  uniprot_id <- unname(unlist(csnk2a_ksr[i,3]))
  site_pos <- unname(unlist(csnk2a_ksr[i,4]))
  
  # Loop for ambiguous psite positions
  
  for (j in 1:length(uniprot_id)) {
    
    pos <- as.numeric(site_pos)
    
    # Substrate sequence
  
    sub_seq <- unlist(getSequence(hs_rp[hs_ids %in% uniprot_id[j]]))
    
    if (length(sub_seq) == 0) {next}
    
    # Phosphosite position
    
    if (pos < 8) {
      
      gap <- paste(rep('_',(8-pos)),collapse='')
      psite <- paste(sub_seq[1:(pos+7)],collapse='')
      psite <- paste(gap,psite,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else if ((pos+7) > length(sub_seq)) {
      
      psite <- paste(sub_seq[(pos-7):(length(sub_seq))],collapse='')
      gap <- paste(rep('_',(pos+7)-length(sub_seq)),collapse='')
      psite <- paste(psite,gap,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else {
      
      psite <- paste(sub_seq[(pos-7):(pos+7)],collapse='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    }
  } 
}

# After removing potential heirarchical sites
bachmann_vec <- psite_vec
p1s <- rapply(strsplit(psite_vec,split=''), function(x) x[9]) == 'S'
p3s <- rapply(strsplit(psite_vec,split=''), function(x) x[11]) == 'S'
psite_no_heira <- psite_vec[!(p1s|p3s)]

ggseqlogo(psite_no_heira,method='probability')

# Construct PWM using basic method

source('match-tm.r')
csnk2_pwm <- makePWM(substr(psite_no_heira,2,14))

# Negative charge scoring function

ED_weight <- function(psite,pwm) {
  
  neg_row <- apply(pwm[rownames(pwm) %in% c('D','E'),],2,sum)
  pwm <- pwm[!rownames(pwm) %in% c('D','E'),]
  pwm <- rbind(pwm,neg_row)
  rownames(pwm)[nrow(pwm)] <- 'DE'
  
  psite <- unlist(strsplit(psite,split=''))
  
  c=0
  min=0
  
  # Ignore the central residue
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    if (aa %in% c('D','E')) {aa <- 'DE'}
    
    # Skip if the amino acid is not D or E
    if (!(aa %in% c('DE'))) {next}
    
    current <- pwm[rownames(pwm) %in% aa,i]
    c = c+current
  
  }
  
  maximum <- sum(pwm[rownames(pwm) == 'DE',][c(1:6,8:13)])
  mss = c/maximum
  
  return(unname(mss))
  
}

```

# Stoichiometry in human (ED charge)

```{r}

library(stringr)

Ishi <- read.xls('Ishihama_stoichiometry.xlsx',sheet=2, stringsAsFactors=FALSE)

## Exclude sites with 0 stoichiometry

Ishi <- Ishi[!Ishi[,8] == 0,]
Ishi_CK2 <- Ishi[grep('CK2',Ishi[,18]),]

# Doubly phosphorylated peptides are potential confounders and should be removed
Ishi_CK2 <- Ishi_CK2[!Ishi_CK2[,3] == '2 Phospho (ST)',]

# Negative charge scoring function
ED_weight <- function(psite,pwm) {
  
  neg_row <- apply(pwm[rownames(pwm) %in% c('D','E'),],2,sum)
  pwm <- pwm[!rownames(pwm) %in% c('D','E'),]
  pwm <- rbind(pwm,neg_row)
  rownames(pwm)[nrow(pwm)] <- 'DE'
  
  psite <- unlist(strsplit(psite,split=''))
  
  c=0
  min=0
  
  # Ignore the central residue
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    if (aa %in% c('D','E')) {aa <- 'DE'}
    
    # Skip if the amino acid is not D or E
    if (!(aa %in% c('DE'))) {next}
    
    current <- pwm[rownames(pwm) %in% aa,i]
    c = c+current
  
  }
  
  maximum <- sum(pwm[rownames(pwm) == 'DE',][c(1:6,8:13)])
  mss = c/maximum
  
  return(unname(mss))
  
}

# Score all non-hierarchical sites with the MSS and then plot

Ishi_CK2_stoi <- Ishi_CK2[,c(8,9)]

# Remove all potential heirarchical sites and T sites
p1s <- rapply(strsplit(Ishi_CK2_stoi[,2],split=''), function(x) x[8]) == 'S'
p3s <- rapply(strsplit(Ishi_CK2_stoi[,2],split=''), function(x) x[10]) == 'S'
p0t <- rapply(strsplit(Ishi_CK2_stoi[,2],split=''), function(x) x[7]) == 'T'
Ishi_CK2_stoi <- Ishi_CK2_stoi[!(p1s|p3s|p0t),]

# Calculate subtrate quality for each phosphosite

mss_vec <- NULL

for (i in 1:nrow(Ishi_CK2_stoi)) {
  
  mss <- NULL
  ck2_site <- Ishi_CK2_stoi[i,2]
  mss <- ED_weight(ck2_site,csnk2_pwm)
  if (length(mss) == 0) {break}
  mss_vec <- c(mss_vec, mss)
  
}

Ishi_CK2_stoi <- cbind(Ishi_CK2_stoi, mss_vec)

```

# Obtain abundance values for the CK2 substrates using PaxDB

```{r}

# Read in the PaxDB table

human_paxdb <- read.table('Human_stoichiometry_PaxDB_integrated.tsv',sep='\t',stringsAsFactors = FALSE)
human_ensembl <- gsub('9606.','',human_paxdb[,2])

# Read in the table that maps between UniProt and Ensembl IDs

uniprot_map <- read.table('full_uniprot_2_paxdb.04.2015.tsv', stringsAsFactors = FALSE)
uniprot_map <- uniprot_map[uniprot_map[,1] == '9606',]
human_ensembl <- gsub('9606.','',human_paxdb[,2])
ensembl_uniprot <- uniprot_map[uniprot_map[,3] %in% human_ensembl,c(2,3)]
ensembl_uniprot[,1] <- rapply(strsplit(ensembl_uniprot[,1],split='\\|'), function(x) x[1])
rownames(ensembl_uniprot) <- NULL

# Retrieve Ensembl IDs of CK2 substrates

Ishi_CK2_ensembl <- Ishi_CK2[Ishi_CK2[,7] %in% ensembl_uniprot[,1],]
Ishi_CK2_ensembl <- Ishi_CK2_ensembl[,c(7,11,8,9)]
colnames(Ishi_CK2_ensembl) <- c('Accession','Position','Stoichiometry','Peptide')

Ishi_CK2_ensembl_ext <- Ishi_CK2_ensembl
ensembl_vec <- NULL
s = 0

for (i in 1:nrow(Ishi_CK2_ensembl)) {
  
  i2 <- i+s
  
  accession <- Ishi_CK2_ensembl[i,1]
  ensembl <- ensembl_uniprot[ensembl_uniprot[,1] %in% accession,2]
  
    # Account for the fact that some accessions map to more than one ensembl ID
  
  if (length(ensembl) != 1) {Ishi_CK2_ensembl_ext <- rbind(Ishi_CK2_ensembl_ext[1:i2,],Ishi_CK2_ensembl_ext[i2,],Ishi_CK2_ensembl_ext[(i2+1):nrow(Ishi_CK2_ensembl_ext),])}
  if (length(ensembl) != 1) {s = s+1}
  
  ensembl_vec <- c(ensembl_vec, ensembl)
  
}

Ishi_CK2_ensembl_ext <- cbind(Ishi_CK2_ensembl_ext, ensembl_vec)
colnames(Ishi_CK2_ensembl_ext) <- c('Accession', 'Position', 'Stoichiometry','Peptide','Ensembl')

# Retrieve abundance data for each CK2 substrate

abundance_vec <- NULL

for (i in 1:nrow(Ishi_CK2_ensembl_ext)) {
  
  ensembl_id <- Ishi_CK2_ensembl_ext[i,5]
  abundance <- human_paxdb[human_ensembl %in% ensembl_id,3]
  abundance_vec <- c(abundance_vec, abundance)
  
}

# Take the log10 of the abundance values

log_abundance_vec <- log(abundance_vec,base=10)
Ishi_abundance <- cbind(Ishi_CK2_ensembl_ext, log_abundance_vec)
colnames(Ishi_abundance) <- c('Accession','Position','Stoichiometry','Peptide','Ensembl','Abundance')

```

# Add substrate quality (relative charge)

```{r}

# Iterate through each site and then calculate the substrate quality (relative charge)

mss_vec <- NULL

for (i in 1:nrow(Ishi_abundance)) {
  
  mss <- NULL
  ck2_site <- Ishi_abundance[i,4]
  mss <- ED_weight(ck2_site,csnk2_pwm)
  if (length(mss) == 0) {break}
  mss_vec <- c(mss_vec, mss)
  
}

Ishi_abundance_charge <- cbind(Ishi_abundance, mss_vec)
colnames(Ishi_abundance_charge) <- c('Accession', 'Position', 'Stoichiometry','Peptide','Ensembl', 'Abundance', 'Charge')

```

# Add interactions from iRefIndex

```{r}

# Read in the iRefindex data for human:

# N.B. This PPI file is too big (2.9 GB) for me to include on GitHub. Please download the PPI file
# for Homo Saoiens from the iRefIndex website: https://irefindex.vib.be/wiki/index.php/iRefIndex

irefindex <- read.table('9606.mitab.27062020.txt', sep='\t', fill=TRUE)

# Retieve PPIs for the two alpha subunits and for the beta subunit

csnk2a1_int <- unique(c(irefindex[grep('P68400',irefindex[,1]),2],irefindex[grep('P68400',irefindex[,2]),1]))
csnk2a2_int <- unique(c(irefindex[grep('P19784',irefindex[,1]),2],irefindex[grep('P19784',irefindex[,2]),1]))
csnk2b_int <- unique(c(irefindex[grep('P67870',irefindex[,1]),2],irefindex[grep('P67870',irefindex[,2]),1])) 

ck2_int <- unique(c(csnk2a1_int,csnk2a2_int,csnk2b_int))

# Check to see if any of the substrates in our data frame interacts with at least one of the CK2 subunits

shared_vec <- NULL

for (i in 1:nrow(Ishi_abundance_charge)) {
  
  print(i)
  
  uniprot <- Ishi_abundance_charge[i,1]
  
  uniprot_int <- unique(c(irefindex[grep(uniprot,irefindex[,1]),2],irefindex[grep(uniprot,irefindex[,2]),1]))
  
  shared_int <- intersect(uniprot_int, c('uniprotkb:P68400','uniprotkb:P19784','uniprotkb:P67870'))
  
  shared_int_len <- length(shared_int)
  
  shared_vec <- c(shared_vec, shared_int_len)
  
}

Ishi_abundance_charge_int <- cbind(Ishi_abundance_charge,shared_vec)
colnames(Ishi_abundance_charge_int) <- c(colnames(Ishi_abundance_charge),'Int_num')

```

# Phosphoacceptor

```{r}

# Check to see if stoichiometry differs between S and T sites

ck2_fasta <- read.fasta('CK2_targets_uniprot.fasta',seqtype = 'AA')
ck2_fasta_names <- rapply(strsplit(names(ck2_fasta),split='\\|'), function(x) x[2])

error_index <- NULL
disorder_res_vec <- NULL

for (i in 1:nrow(Ishi_abundance_charge_int)) {
  
  accession <- Ishi_abundance_charge_int[i,1]
  pos <- Ishi_abundance_charge_int[i,2]
  
  if (accession=='Q5SNT6') {
    
    accession <- 'Q641Q2'
    pos <- 1054
    
  }
  
  if (accession=='Q8N9T8') {
    
    accession <- 'Q8N9T8'
    pos <- 171
    
  }
  
  target_seq <- ck2_fasta[ck2_fasta_names %in% accession]
  target_seq <- unlist(getSequence(target_seq))
  disorder_res <- target_seq[pos]
  
  if (!disorder_res %in% c('S','T','Y') | length(disorder_res) == 0) {
    stop()
  }
  
  disorder_res_vec <- c(disorder_res_vec, disorder_res)
  
}

```

# Disorder content prediction (for new version use AF2 instead of SPOT Disorder-Single)

```{r}

# We will predict protein disorder by using an AF2-based method where the smoothened surface
# accessibility within a sequence window is used as a proxy for protein disorder.
# The approach is described here, and it is where we source the accessibility data:
# https://github.com/normandavey/ProcessedAlphafold

# For proteins without protein structures we use the SPOT-Disorder-Single predictor

Ishi_disorder_scaled <- readRDS('Ishi_disorder_scaled.rds')
AF2_accessibility <- readLines('9606.accessibility_windowed_scores.15.tdt')

AF2_vec <- NULL

count = 0

for (i in 1:nrow(Ishi_abundance_charge_int)) {
  
  print(i)
  
  accession <- Ishi_abundance_charge_int[i,1]
  pos <- Ishi_abundance_charge_int[i,2]
  
  if (accession=='Q5SNT6') {
    
    accession <- 'Q641Q2'
    pos <- 1054
    
  }
  
  if (accession=='Q8N9T8') {
    
    accession <- 'Q8N9T8'
    pos <- 171
    
  }
  
  grep_dex <- grep(accession,AF2_accessibility)
  
  # Use SPOT-disorder-single if there is no AF2 structure
  
  if (length(grep_dex) == 0) {disorder_pred <- Ishi_disorder_scaled[i,13]; AF2_vec <- c(AF2_vec,disorder_pred); 
  count = count+1; next}
  
  disorder_pred <- AF2_accessibility[grep_dex]
  disorder_pred <- rapply(strsplit(disorder_pred,split='\t'), function(x) x[2])
  disorder_pred <- unlist(strsplit(disorder_pred,split=','))
  disorder_pred <- disorder_pred[pos]
  
  AF2_vec <- c(AF2_vec,disorder_pred)
  
}

AF2_vec <- as.numeric(AF2_vec)

#####

# Use 0.55 disorder threshold for binary classification of protein disorder
# https://github.com/normandavey/ProcessedAlphafold

disorder_class_vec <- rep('D',length(AF2_vec))
disorder_class_vec[AF2_vec <= 0.55] <- 'O'

Ishi_disorder <- cbind(Ishi_abundance_charge_int,disorder_res_vec,AF2_vec,disorder_class_vec)
colnames(Ishi_disorder) <- c(colnames(Ishi_abundance_charge_int),'Phosphoacceptor','Disorder_score','Disorder_class')

```

# Add subcellular localisation

```{r}

# Subcellular locations for each CK2 substrate were scraped from UniProt

CK2_uniprot_scrape <- readLines('CK2_UniProt_raw.txt')

CK2_target_SC_vec <- NULL

# Parse the raw UniProt entries to the retrieve the subcellular localisation information

for (i in 1:nrow(Ishi_disorder)) {
     
    accession <- Ishi_disorder[i,1]
    
    CK2_targets_start_AC <- grep('AC  ',CK2_uniprot_scrape)
    CK2_targets_start_accession <- grep(accession,CK2_uniprot_scrape)
    
    start <- intersect(CK2_targets_start_AC,CK2_targets_start_accession)
    if (length(start) >1) {stop('Start problem')}
    
    end_grep <- grep('FT   ',CK2_uniprot_scrape)
    end_grep <- end_grep[which((end_grep-start) > 0)]
    end <- end_grep[which((end_grep-start) == min(end_grep-start))]
    
    #####
    
    CK2_target_scrape <- CK2_uniprot_scrape[start:end]
    
    CK2_target_GO <- grep('GO;',CK2_target_scrape)
    CK2_target_C <- grep('C:',CK2_target_scrape)  
    
    CK2_target_GO_C <- intersect(CK2_target_GO,CK2_target_C)
    
    CK2_target_GO_C_text <- CK2_target_scrape[CK2_target_GO_C]
    
    if (length(CK2_target_GO_C_text) == 0) {location <- 'Unknown'; CK2_target_SC_vec <- c(CK2_target_SC_vec,location); next}
    
    GO_compartments <- rapply(strsplit(rapply(strsplit(CK2_target_GO_C_text,split=' C:'), function(x) x[2]), split=';'), function(x) x[1])
    
    #####
    
    # Assign the location to nucleus (N), cytoplasm (C), or mitochondria (M) depending upon matches to the corresponding GO terms
    
    N <- 0
    C <- 0
    M <- 0
    
    nuclear_terms <- c('nucleus','nucleoplasm','nuclear envelope', 'nuclear speck','ribonucleoprotein complex','nucleolus', 'chromosome')
    cyto_terms <- c('cytoplasm','cytosol')
    mito_terms <- c('mitochondria', 'mitochondrion', 'mitochondrial inner membrane', 'mitochondrial outer membrane', 'mitochondrial membrane')
    
    if (length(intersect(GO_compartments,nuclear_terms)) > 0) {N <- 1}
    if (length(intersect(GO_compartments,cyto_terms)) > 0) {C <- 1}
    if (length(intersect(GO_compartments,mito_terms)) > 0) {M <- 1}
    
    if (N == 1 & C == 1 & M == 1) {location <- 'NCM'}
    if (N == 1 & C == 0 & M == 1) {location <- 'NM'}
    if (N == 1 & C == 1 & M == 0) {location <- 'NC'}
    if (N == 0 & C == 1 & M == 1) {location <- 'CM'}
    if (N == 1 & C == 0 & M == 0) {location <- 'N'}
    if (N == 0 & C == 1 & M == 0) {location <- 'C'}
    if (N == 0 & C == 0 & M == 1) {location <- 'M'}
    if (N == 0 & C == 0 & M == 0) {location <- 'None'}
    
    CK2_target_SC_vec <- c(CK2_target_SC_vec,location)
    
}

# Group together infrequent labels to maximise power

CK2_target_SC_vec <- gsub('^CM$','Other',CK2_target_SC_vec)
CK2_target_SC_vec <- gsub('^NCM$','Other',CK2_target_SC_vec)
CK2_target_SC_vec <- gsub('^NM$','Other',CK2_target_SC_vec)
CK2_target_SC_vec <- gsub('^M$','Other',CK2_target_SC_vec)
CK2_target_SC_vec <- gsub('^None$','Other',CK2_target_SC_vec)
CK2_target_SC_vec <- gsub('^Unknown$','Other',CK2_target_SC_vec)

# Add to data frame

Ishi_disorder_location <- cbind(Ishi_disorder, CK2_target_SC_vec)
colnames(Ishi_disorder_location) <- c(colnames(Ishi_disorder), 'Location')

```

# Save the data frame

```{r}

saveRDS(Ishi_disorder_location,'Ishi_disorder_CK2_github.rds')

```
