---
title: "Supplementary figure 5C"
author: "David Bradley"
date: "06/12/2023"
output: html_document
---

# GGplot font

```{r}

# https://stackoverflow.com/questions/27689222/changing-fonts-for-graphs-in-r

library(extrafont)

font_import()
loadfonts()       #Register fonts for Windows bitmap output
fonts() 

```

# Retrieve CK2 substrates and set up scoring function for calculating substrate quality

```{r}

library(readr)
library(seqinr)
library(ggseqlogo)
library(gdata)
library(stringr)

# Data comes from Bachmann et al., 2019 (https://www.biorxiv.org/content/10.1101/822668v3)

prot_mapper <- read_csv('export.csv')

# Kinases only

prot_mapper <- prot_mapper[unlist(prot_mapper[,5]) == TRUE,]

# Retrieve upstream kinase, substrate, and substrate position

ksr <- prot_mapper[,c(3,4,6,9,10)]

# We exclude KSRs that are supported only by the text-mining tools

curated <- c('psp','signor','pid','bel','reactome')

source_list <- strsplit(unlist(ksr[,5]),split=',')

number_curated <- lapply(source_list, function(x) length(intersect(x,curated)))

ksr <- ksr[which(number_curated > 0),]

# Correct formatting problems

for (i in 1:nrow(ksr)) {
 
  if(is.na(ksr[i,2])) {ksr[i,2] <- ksr[i,1]} 
  
}

# Retrieve CSNK2A1 and CSNK2A2 substrates

csnk2a1_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A1') == TRUE),]
csnk2a2_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A2') == TRUE),]
csnk2a_ksr <- unique(rbind(csnk2a1_ksr,csnk2a2_ksr))

# Human proteome

hs_rp <- read.fasta('hs_proteome_uniprot.fasta', seqtype='AA')
hs_ids <- rapply(strsplit(names(hs_rp),split='\\|'),function(x) x[2])

# Retrieve the sequences

psite_vec <- NULL
index_vec <- NULL

for (i in 1:nrow(csnk2a_ksr)) {
  
  print(i)
  
  # Retrieve the first element only
  uniprot_id <- unname(unlist(csnk2a_ksr[i,3]))
  site_pos <- unname(unlist(csnk2a_ksr[i,4]))
  
  # Loop for ambiguous psite positions
  
  for (j in 1:length(uniprot_id)) {
    
    pos <- as.numeric(site_pos)
    
    # Substrate sequence
  
    sub_seq <- unlist(getSequence(hs_rp[hs_ids %in% uniprot_id[j]]))
    
    if (length(sub_seq) == 0) {next}
    
    # Phosphosite position
    
    if (pos < 8) {
      
      gap <- paste(rep('_',(8-pos)),collapse='')
      psite <- paste(sub_seq[1:(pos+7)],collapse='')
      psite <- paste(gap,psite,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else if ((pos+7) > length(sub_seq)) {
      
      psite <- paste(sub_seq[(pos-7):(length(sub_seq))],collapse='')
      gap <- paste(rep('_',(pos+7)-length(sub_seq)),collapse='')
      psite <- paste(psite,gap,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else {
      
      psite <- paste(sub_seq[(pos-7):(pos+7)],collapse='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    }
  } 
}

# Remove potential heirarchical sites
bachmann_vec <- psite_vec
p1s <- rapply(strsplit(psite_vec,split=''), function(x) x[9]) == 'S'
p3s <- rapply(strsplit(psite_vec,split=''), function(x) x[11]) == 'S'
psite_no_heira <- psite_vec[!(p1s|p3s)]

# Retrieve the corresponding database records

csnk2a_ksr_no_heira <- csnk2a_ksr[index_vec[!(p1s|p3s)],]

csnk2a_ksr_no_heira <- cbind(csnk2a_ksr_no_heira,psite_no_heira)

# Generate the CK2 PWM

source('match-tm.r')

csnk2_pwm <- makePWM(substr(psite_no_heira,2,14))

# Function for calculating the CK2 substrate quality

ED_weight <- function(psite,pwm) {
  
  neg_row <- apply(pwm[rownames(pwm) %in% c('D','E'),],2,sum)
  pwm <- pwm[!rownames(pwm) %in% c('D','E'),]
  pwm <- rbind(pwm,neg_row)
  rownames(pwm)[nrow(pwm)] <- 'DE'
  
  psite <- unlist(strsplit(psite,split=''))
  
  c=0
  min=0
  
  # Ignore the central residue
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    if (aa %in% c('D','E')) {aa <- 'DE'}
    # Skip if the amino acid is not D or E
    if (!(aa %in% c('DE'))) {next}
    
    current <- pwm[rownames(pwm) %in% aa,i]
    c = c+current
  
  }
  
  maximum <- sum(apply(pwm[,c(1:6,8:13)],2,max))
  mss = c/maximum
  
  return(unname(mss))
  
}

```

# Check for cases where the known CK2 substrate has more than one upstream kinases that is recorded in ProtMapper

```{r}

### Correct for duplicated records

csnk2a_ksr_no_heira <- csnk2a_ksr_no_heira[!duplicated(csnk2a_ksr_no_heira[c(3,4)]),]

library(readr)
library(seqinr)
library(ggseqlogo)
library(gdata)

# Data comes from Bachmann et al., 2019 (https://www.biorxiv.org/content/10.1101/822668v3)

prot_mapper <- read_csv('export.csv')

# Kinases only 

prot_mapper <- prot_mapper[prot_mapper[,5] == TRUE,]

# Retrieve upstream kinase, substrate, and substrate position

ksr <- prot_mapper[,c(4,6,9)]

ksr_string <- paste(unlist(ksr[,2]),unlist(ksr[,3]),sep='_')

# Now retrieve the phosphosites from the human cell cycle phosphoproteome study (Olsen et al., 2010)

library(stringr)
library(readxl)
library(seqinr)
library(gdata)

# Olsen phosphoproteome

# Read in the file

Olsen_S2 <- read.xls('Olsen_2010_S2.xls',sheet=1, stringsAsFactors=FALSE)

# Class I sites only

Olsen_S2 <- Olsen_S2[grep('YES',Olsen_S2[,5]),]

# Single phosphosite only

Olsen_S2 <- Olsen_S2[Olsen_S2[,16] == 1,]

# Retrieve sequence window

Olsen_sites <- Olsen_S2[,18]

# Known CSNK2 substrates
csnk2_sub <- unname(unlist(csnk2a_ksr_no_heira[,3]))

# Retrieve the Olsen UniProt accessions
Olsen_S2_UP <- strsplit(Olsen_S2[,2],split=';')

# Match the cell cycle phosphosites to the known CK2 substrates

col1 <- unlist(Olsen_S2_UP)[unlist(lapply(Olsen_S2_UP, function(x) x %in% csnk2_sub))]

i_vec <- NULL

for (i in 1:length(Olsen_S2_UP)) {
  
 i_vec <- c(i_vec, rep(i,length(Olsen_S2_UP[[i]])))
  
}

pos <- as.numeric(Olsen_S2[i_vec,3])

col2 <- pos[unlist(lapply(Olsen_S2_UP, function(x) x %in% csnk2_sub))]

col1_col2 <- cbind(col1,col2)

match_dex <- match(paste(unlist(csnk2a_ksr_no_heira[,3]),unlist(csnk2a_ksr_no_heira[,4])),paste(col1,col2))

olsen_ck2_sub <- csnk2a_ksr_no_heira[!is.na(match_dex),]

# Retrieve the corresponding rows in the original supplementary table

grep_vec <- NULL

for (i in 1:nrow(olsen_ck2_sub)) {
  
  sub <- unlist(unname(olsen_ck2_sub[i,3]))
  pos <- unlist(unname(olsen_ck2_sub[i,4]))
  
  grep1 <- grep(sub,Olsen_S2[,2])
  grep2 <- grep(paste0("^",pos,"$"),Olsen_S2[,3])
  
  grep <- intersect(grep1,grep2)
  
  if (length(grep) > 1) {stop()}
    
  grep_vec <- c(grep_vec, grep)
  
}

Olsen_S2_ck2 <- Olsen_S2[grep_vec,]

# Retrieve SILAC ratios for the six stages tested

Olsen_S2_ck2_ratios <- Olsen_S2_ck2[,6:11]

# Retrieve the flanking sequences of the CK2 sites

ck2_sites <- Olsen_S2_ck2[,18]

# Calculate the substrate quality of the CK2 sites

ED_weight <- function(psite,pwm) {
  
  neg_row <- apply(pwm[rownames(pwm) %in% c('D','E'),],2,sum)
  pwm <- pwm[!rownames(pwm) %in% c('D','E'),]
  pwm <- rbind(pwm,neg_row)
  rownames(pwm)[nrow(pwm)] <- 'DE'
  
  psite <- unlist(strsplit(psite,split=''))
  
  c=0
  min=0
  
  # Ignore the central residue
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    if (aa %in% c('D','E')) {aa <- 'DE'}
    # Skip if the amino acid is not D or E
    if (!(aa %in% c('DE'))) {next}
    
    current <- pwm[rownames(pwm) %in% aa,i]
    c = c+current
  
  }
  
  maximum <- sum(apply(pwm[,c(1:6,8:13)],2,max))
  mss = c/maximum
  
  return(unname(mss))
  
}
csnk2_pwm <- makePWM(substr(psite_no_heira,2,14))

mss_vec <- NULL

for (i in 1:length(ck2_sites)) {
  
  mss <- NULL
  ck2_site <- ck2_sites[i]
  mss <- ED_weight(ck2_site,csnk2_pwm)
  if (length(mss) == 0) {break}
  mss_vec <- c(mss_vec, mss)
  
}

# Bind the SQ to the Prot Mapper data

olsen_ck2_sub <- cbind(olsen_ck2_sub, mss_vec)

# For each site, check to see if there are more upstream kinases besides from CK2

olsen_ck2_string <- paste(olsen_ck2_sub[,3],olsen_ck2_sub[,4],sep='_')

length_vec <- NULL

for (i in 1:length(olsen_ck2_string)) {
  
  ck2_string <- olsen_ck2_string[i]
  len <- length(ksr_string[ksr_string %in% ck2_string])
  length_vec <- c(length_vec, len)
}

olsen_ck2_sub <- cbind(olsen_ck2_sub, length_vec)
olsen_ck2_sub[olsen_ck2_sub[,7] < 0.4,8]
olsen_ck2_sub[olsen_ck2_sub[,7] > 0.4 & olsen_ck2_sub[,3] < 0.6,8]
olsen_ck2_sub[olsen_ck2_sub[,7] > 0.6,8]

# set up the data frame for the box plot

options(warn=1)

ck2_cov_list2 <- NULL
col1 <- NULL
col2 <- NULL

for (i in 0:3) {
  
  lower <- 0.2*i
  upper <- lower+0.2
  print(lower)
  print(upper)
  
  ck2_optimal  <- ck2_sites[mss_vec > lower & mss_vec <= upper]
  ck2_optimal_ratios <- olsen_ck2_sub[mss_vec > lower & mss_vec <= upper,]
  
  # Filter out rows with any missing values
  
  na_count <- apply(ck2_optimal_ratios, 1, function(x) length(which(is.na(x))))
  ck2_optimal_ratios <- ck2_optimal_ratios[na_count < 1,]
  
  ck2_cov <- ck2_optimal_ratios[,8]
  
  ck2_cov <- ck2_cov[!is.na(ck2_cov)]
  ck2_cov <- ck2_cov[!is.infinite(ck2_cov)]
  ck2_cov <- unname(ck2_cov)
  
  ck2_cov_list2 <- c(ck2_cov_list2, list(ck2_cov))
  
  col1 <- c(col1, ck2_cov)
  col2 <- c(col2, rep(paste(lower,'-',upper,sep=''),length(ck2_cov)))
  
}

mss_data <- data.frame(col1,col2)
mss_data[,2] <- as.factor(mss_data[,2])
colnames(mss_data) <- c('Stoichiometry','Positions')

# use ggplot to plot the relationship between substrate quality and the number of additional upstream kinases

p <- ggplot(mss_data, aes(x=Positions, y=Stoichiometry)) + geom_boxplot(fill='white',color=colors()[160], lwd=1.1, notch=FALSE,fatten=0.0, width=0.5, outlier.shape=NA) 
dat <- ggplot_build(p)$data[[1]]
p <- p + geom_segment(data=dat, aes(x=xmin, xend=xmax, y=middle, yend=middle), colour=colors()[160], size=2.2)
p <- p + geom_jitter(width = 0.3, height=0, color=colors()[132],size=1.5)


p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15))

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")) )

p <- p + ylab("No. of additional upstream kinases") + xlab("Substrate quality") + ggtitle('')
p <- p + theme(axis.title.y = element_text(vjust=-1.5))
p <- p+theme(axis.text.x=element_text(size=6),axis.text.y=element_text(size=9),axis.title.x=element_text(size=8.5,face="plain"),axis.title.y=element_text(size=8.5,face="plain"),plot.title=element_text(size=18,face='plain'))
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(panel.border= element_blank())
p <- p + theme(legend.position = "none") #
p <- p + theme(plot.margin = unit(c(-0.5,0.5,0,0), "cm")) # t, r, b, l
p <- p+coord_cartesian(ylim = c(0, 7))
p <- p + theme(axis.text.x = element_text(angle = 30, vjust = 0.9, hjust=1))
p <- p + theme(axis.title.x = element_text(vjust=4.0))
p <- p+geom_text(data=data.frame(), aes(x=c(1:length(ck2_cov_list2)), y=rep(7,length(ck2_cov_list2)), label=rapply(ck2_cov_list2, function(x) length(x))), col='black', size=2.5)
p <- p + theme(plot.margin = unit(c(0,0.75,0.0,0), "cm")) # t, r, b, l
p

supp_fig_5b_mss_data <- mss_data
supp_fig_5b <- p

ggsave(file='Supplementary_figure_5B.pdf', plot=p, width=3.8, height=4.8)

```

# For the known CK2 substrates, check to see the number of predicted upstream kinases according to Johnson et al., 2023

```{r}

olsen_ck2_sub <- olsen_ck2_sub

# Load in the data from Johnson et al., 2023

johnson <- readRDS('ck2_sites_Johnson_2023.rds')

# Find the Olsen sites that also have predicted substrates

olsen_acc_pos <- paste(olsen_ck2_sub[,3],olsen_ck2_sub[,4],sep='_')
johnson_acc_pos <- paste(johnson[,3],johnson[,4],sep='_')
johnson_promiscuity <- johnson[match(olsen_acc_pos,johnson_acc_pos),8]
johnson_kinases <- johnson[match(olsen_acc_pos,johnson_acc_pos),14]

# Add Johnson promiscuity to the Olsen sites

olsen_ck2_sub <- data.frame(olsen_ck2_sub,johnson_promiscuity)
olsen_ck2_sub <- data.frame(olsen_ck2_sub,johnson_kinases)

# Remove sites that have 0 substrate quality

olsen_ck2_sub <- olsen_ck2_sub[olsen_ck2_sub[,7] > 0,]

# Predicted CK2A1 sites

ck2a1_dex <- grep('CK2A1',olsen_ck2_sub[,10])

# Predicted CK2A2 sites

ck2a2_dex <- grep('CK2A2',olsen_ck2_sub[,10])

# Remove CK2a1 from the tally

olsen_ck2_sub[ck2a1_dex,9] <- olsen_ck2_sub[ck2a1_dex,9]-1

# Remove CK2a1 from the tally

olsen_ck2_sub[ck2a2_dex,9] <- olsen_ck2_sub[ck2a2_dex,9]-1

```

# Generate the GGplot

```{r}

ck2_sites_CC_johnson <- olsen_ck2_sub

p <- ggplot(ck2_sites_CC_johnson,aes(x=mss_vec,y=johnson_promiscuity)) + geom_point(size=1) # Warning message is for sites with NA promiscuity 
p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15))
p <- p+theme(panel.border = element_blank())
p <- p + coord_cartesian(xlim=c(0,1))
p <- p + xlab('Substrate quality') + ylab('Substrate promiscuity (predicted)') + ggtitle('Known CK2 targets')

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")) )

p <- p+theme(axis.text.x=element_text(size=9.5),axis.text.y=element_text(size=10.5),axis.title.x=element_text(size=11.5,face="plain"),axis.title.y=element_text(size=10.75,face="plain"),plot.title=element_text(size=11,face='bold'))
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p+theme(plot.title = element_text(vjust = -1))

supp_fig_5c <- p

```




