---
title: "Figure 3b 3c"
author: "David Bradley"
date: "1/1/2021"
output: html_document
---

# GGplot font

```{r}

# https://stackoverflow.com/questions/27689222/changing-fonts-for-graphs-in-r

library(extrafont)

font_import()
loadfonts()   #Register fonts for Windows bitmap output
fonts() 

```

# Retrieve the known substrates of human CK2

```{r}

library(stringr)
library(readr)
library(seqinr)
library(ggseqlogo)
library(gdata)

# Data comes from Bachmann et al., 2019 (https://www.biorxiv.org/content/10.1101/822668v3)

prot_mapper <- read_csv('export.csv')

# Kinases only

prot_mapper <- prot_mapper[unlist(prot_mapper[,5]) == TRUE,]

# Retrieve upstream kinase, substrate, and substrate position

ksr <- prot_mapper[,c(3,4,6,9,10)]

# We exclude KSRs that are supported only by the text-mining tools

curated <- c('psp','signor','pid','bel','reactome')

source_list <- strsplit(unlist(ksr[,5]),split=',')

number_curated <- lapply(source_list, function(x) length(intersect(x,curated)))

ksr <- ksr[which(number_curated > 0),]

# Correct formatting problems

for (i in 1:nrow(ksr)) {
 
  if(is.na(ksr[i,2])) {ksr[i,2] <- ksr[i,1]} 
  
}

# Retrieve CSNK2A1 and CSNK2A2 substrates

csnk2a1_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A1') == TRUE),]
csnk2a2_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A2') == TRUE),]
csnk2a_ksr <- unique(rbind(csnk2a1_ksr,csnk2a2_ksr)) 

hs_rp <- read.fasta('hs_proteome_uniprot.fasta', seqtype='AA')
hs_ids <- rapply(strsplit(names(hs_rp),split='\\|'),function(x) x[2])

# Retrieve the flanking sequences of the phosphosites

psite_vec <- NULL
index_vec <- NULL

for (i in 1:nrow(csnk2a_ksr)) {
  
  print(i)
  
  # Retrieve the first element only
  uniprot_id <- unname(unlist(csnk2a_ksr[i,3]))
  site_pos <- unname(unlist(csnk2a_ksr[i,4]))
  
  # Loop for ambiguous psite positions
  
  for (j in 1:length(uniprot_id)) {
    
    pos <- as.numeric(site_pos)
    
    # Substrate sequence
  
    sub_seq <- unlist(getSequence(hs_rp[hs_ids %in% uniprot_id[j]]))
    
    if (length(sub_seq) == 0) {next}
    
    # Phosphosite position
    
    if (pos < 8) {
      
      gap <- paste(rep('_',(8-pos)),collapse='')
      psite <- paste(sub_seq[1:(pos+7)],collapse='')
      psite <- paste(gap,psite,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else if ((pos+7) > length(sub_seq)) {
      
      psite <- paste(sub_seq[(pos-7):(length(sub_seq))],collapse='')
      gap <- paste(rep('_',(pos+7)-length(sub_seq)),collapse='')
      psite <- paste(psite,gap,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else {
      
      psite <- paste(sub_seq[(pos-7):(pos+7)],collapse='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    }
  } 
  
}

# Remove potential hierarchical sites
bachmann_vec <- psite_vec
p1s <- rapply(strsplit(psite_vec,split=''), function(x) x[9]) == 'S'
p3s <- rapply(strsplit(psite_vec,split=''), function(x) x[11]) == 'S'
psite_no_heira <- psite_vec[!(p1s|p3s)]

############################################################

# Restrict in the range of -6 to +6

psite_no_heira <- substr(psite_no_heira,2,14)

# Remove any sites with a gap in them

psite_no_heira <- psite_no_heira[-grep('_',psite_no_heira)]

```

# The expected frequencies for aspartate and glutamate in the proteome can be obtained roughly by
# multiplying the nucleotide frequencies for each respective codon encoding aspartate and glutanate

```{r}

# http://www.tiem.utk.edu/~gross/bioed/webmodules/aminoacid.htm
# https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0148174#sec002 (PMID: 26849571)

A = 0.262
T = 0.219
C = 0.257
G = 0.262

D <- 0.262*0.262*0.219+0.262*0.262*0.257
E <- 0.262*0.262*0.262+0.262*0.262*0.262

# Calculate observed background D+E frequency from the human proteome derived from UniProt

library(seqinr)

fasta <- read.fasta('hs_proteome_uniprot.fasta')

table(unname(unlist(fasta)))/sum(table(unname(unlist(fasta))))

# Observed frequency from the human proteome

obs_DE <- 4.797186e-02+7.122659e-02

```

### Calculate the empirical distribution of D+Es of known CK2 substrates. As explained in the paper,
### we restrict this analysis only to the D and Es over and above those found at position +1 and +3

```{r}

psite_no_heira <- unique(psite_no_heira)
p1_dex <- substr(psite_no_heira,8,8) %in% c('D','E')
p3_dex <- substr(psite_no_heira,10,10) %in% c('D','E')
ck2_minimal <- psite_no_heira[p1_dex & p3_dex]

# Calculate the empirical D+E distribution

DE_count <- str_count(ck2_minimal, pattern='E')+str_count(ck2_minimal, pattern='D')

DE_count_marginal <- table(DE_count)/sum(table(DE_count))

# Account for the absence of any phosphosites with 12 Ds + Es

DE_count_marginal <- c(DE_count_marginal, 0)
names(DE_count_marginal)[c(11)] <- c(12)

# Subtract -2 from the scores because we are counting the D and Es over and above those found at +1 and +3

names(DE_count_marginal) <- as.numeric(names(DE_count_marginal))-2

```

# Generated the expected distribution of D+E under neutrality but this time use disorder-adjusted expected frequency for D+E

```{r}

library(ggplot2)

# Read in the disorder-adjusted expected frequencies of amino acids in phosphosites

Bachmann_bg <- readRDS('bachmann_disorder_bg_weights.rds')

disorder_DE <- Bachmann_bg[3]+Bachmann_bg[4]
disorder_DE <- unname(disorder_DE)

prob_vec <- NULL

for (i in 0:10) {
  
  prob <- dbinom(x=i,size=10,prob=disorder_DE)
  print(prob)
  prob_vec <- c(prob_vec, prob)
  
}

# GGplot line plot of the theoretical distribution under neutrality (orange) and observed distribution (black). The
# color of the theoretical line matches the color for the corresponding KL divergence when comparing the empirical
# and theoretical distributions.

col1 <- c(0:10,0:10)
col2 <- c(prob_vec,DE_count_marginal)
col3 <- c(rep('Model',11),rep('Empirical',11))
df <- data.frame(col1,col2,col3)
colnames(df) <- c('Pos','Prob','Group')

p <- ggplot(data=df, aes(x=Pos, y=Prob, group=Group)) +
  geom_line(aes(color=Group))+
  geom_point(aes(color=Group))
p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15))
p <- p + xlab('') + ylab('Relative frequency') + ggtitle('')
p <- p + scale_x_continuous(breaks=c(seq(1,14,2)))
p <- p + scale_color_manual(values=c('black','#F97C5DFF'))
p <- p + coord_cartesian(ylim = c(0, 0.4))
p <- p + theme(legend.title = element_blank())

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")) )

p <- p + theme(plot.margin = unit(c(0,0.3,0.0,1), "cm")) # t, r, b, l
p <- p+theme(axis.text=element_text(size=11),axis.title.x=element_text(size=13,face="plain"),axis.title.y=element_text(size=11,face="plain"))
p <- p + theme(legend.title=element_text(size=0),legend.text=element_text(size=11))
p <- p+theme(legend.position="none")
p

p <- p+geom_text(x=3.92, y=0.275, label='KL = 2.2',size=4.5,fontface='bold',col='#F97C5DFF',parse=F, check_overlap=T)

g2 <- p

ggsave(file='Figure_3b.pdf', plot=p, width=6.0, height=4.0)

```

# Chi-squared test for no. of sites with >=3 additional charges

```{r}

psite_no_heira <- unique(psite_no_heira)
p1_dex <- substr(psite_no_heira,8,8) %in% c('D','E')
p3_dex <- substr(psite_no_heira,10,10) %in% c('D','E')
ck2_minimal <- psite_no_heira[p1_dex & p3_dex]

# Empirical

DE_count <- str_count(ck2_minimal, pattern='E')+str_count(ck2_minimal, pattern='D')
DE_count <- DE_count-2

emp_3_or_more <- length(which(DE_count >= 3))
emp_2_or_less <- length(which(DE_count < 3))

# Theoretical

DE_count_theoretical <- round(prob_vec*length(ck2_minimal))
names(DE_count_theoretical) <- 0:10
theory_3_or_more <- sum(DE_count_theoretical[4:11])
theory_2_or_less <- sum(DE_count_theoretical[1:3])

theoretical_vec <- c(rep(0,49),rep(1,77),rep(2,55),rep(3,24),rep(4,7),rep(5,1))

# Contingency table

ck2_mat <- matrix(c(emp_2_or_less,emp_3_or_more,theory_2_or_less,theory_3_or_more),ncol=2,byrow=T)
rownames(ck2_mat) <- c('Emp','Thr')
chisq.test(ck2_mat)$p.value

```

# Now calculate the output of the model under selection

```{r}

# Effective population size for human (from https://doi.org/10.1016/j.jmb.2020.02.006)

Ne <- 21091

# Calculate the probability vector under neutrality

prob_vec <- NULL

for (i in 0:10) {
  
  prob <- dbinom(x=i,size=10,prob=disorder_DE)
  
  print(prob)
  
  prob_vec <- c(prob_vec, prob)
  
}

# Now model the effect of selection

theta <- 10 # specified fitness optimum (scale parameter)
m <- 0:10 # number of D/Es

omega <- 480 # relative strength of selection between genotypic states (shape parameter)

numerator <- -(m-theta)^2 # numerator of the exponent of the fitness function
denominator <- (2*omega^2) # denominator of the exponent of the fitness function
exponent <- numerator / denominator
fitness <- exp(exponent)

s = fitness

# Normalise

sm <- s-max(s) # sm is the relative fitness

sel_mod <- exp(2*Ne*sm) # in essence this is the factor by which the probabilities under neutrality (prob_vec) are scaled by selection

## scale the probabilities under neutrality (by selection) and then normalise so that everything sums to 1

prob_vec <- (prob_vec*sel_mod)/sum(prob_vec*sel_mod)

## GGplot

col1 <- c(0:10,0:10)
col2 <- c(prob_vec,DE_count_marginal)
col3 <- c(rep('Model',11),rep('Empirical',11))
df <- data.frame(col1,col2,col3)
colnames(df) <- c('Pos','Prob','Group')

p <- ggplot(data=df, aes(x=Pos, y=Prob, group=Group)) +
  geom_line(aes(color=Group))+
  geom_point(aes(color=Group))
p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15))
p <- p + xlab('no. of D + E') + ylab('') + ggtitle('')
p <- p + scale_x_continuous(breaks=c(seq(1,14,2)))
p <- p + scale_color_manual(values=c('black','#FED395FF'))
p <- p + coord_cartesian(ylim = c(0, 0.4))
p <- p + theme(legend.title = element_blank())
p <- p+theme(legend.position="none")

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")) )

p <- p + theme(plot.margin = unit(c(0,1.15,0.0,0), "cm")) # t, r, b, l
p <- p+theme(axis.text=element_text(size=11),axis.title.x=element_text(size=13,face="plain"),axis.title.y=element_text(size=11,face="plain"))
p <- p + theme(legend.title=element_text(size=0),legend.text=element_text(size=11))
#p <- p + theme(plot.margin = unit(c(0,0,0.0,-0.01), "cm")) # t, r, b, l
p

p <- p+geom_text(x=6.2, y=0.277, label='KL = 0.17',size=4.5,fontface='bold',col='#FED395FF',parse=F, check_overlap=T)

g3 <- p

```


