---
title: "Supplementary figure 3"
output: html_document
---

```{r}

# https://stackoverflow.com/questions/27689222/changing-fonts-for-graphs-in-r

library(extrafont)

font_import()
loadfonts()       #Register fonts for Windows bitmap output
fonts() 

```


# Read in data (independent WT peptides)

```{r}

library(gdata)
library(readxl)
library(seqinr)
library(stringr)

ck2_peptide_library <- read_xlsx('CK2_peptides_library.xlsx', skip=0, col_names=TRUE, sheet=1)

# Keep the WT and mutant peptide data but discard the data pertaining to orthologues
ck2_peptide_library <- ck2_peptide_library[476:nrow(ck2_peptide_library),]
ck2_peptide_library <- as.data.frame(ck2_peptide_library)

# WT peptide only
MSS <- as.numeric(ck2_peptide_library[,4][!is.na(ck2_peptide_library[,2])])
phosphosignal <- as.numeric(ck2_peptide_library[,6][!is.na(ck2_peptide_library[,2])])
ck2_sites <- ck2_peptide_library[,1][!is.na(ck2_peptide_library[,2])]

# Exclude 0 values as we need to take logarithms of the phosphorylation signal
MSS <- MSS[!phosphosignal == 0]
ck2_sites <- ck2_sites[!phosphosignal == 0]
phosphosignal <- phosphosignal[!phosphosignal == 0]

# Take logs and then plot
phosphosignal <- log(phosphosignal,base=10)

# Assign WT labels to the phosphorylation signal and CK2 peptides

phosphosignal_wt <- phosphosignal
ck2_sites_wt <- ck2_sites

```

# Read in data for the full peptide set (WT+mutants)

```{r}

ck2_peptide_library <- read_xlsx('CK2_peptides_library.xlsx', skip=0, col_names=TRUE, sheet=1)

# Keep the WT and mutant peptide data but discard the data pertaining to orthologues
ck2_peptide_library <- ck2_peptide_library[476:nrow(ck2_peptide_library),]
ck2_peptide_library <- as.data.frame(ck2_peptide_library)

# WT and mutant peptides
MSS <- as.numeric(ck2_peptide_library[,4])
phosphosignal <- as.numeric(ck2_peptide_library[,6])
ck2_sites <- ck2_peptide_library[,1]

# Remove NAs
phosphosignal <- phosphosignal[!is.na(MSS)]
ck2_sites <- ck2_sites[!is.na(MSS)]
MSS <- MSS[!is.na(MSS)]

# Exclude 0 values as we need to take logarithms of the phosphorylation signal
MSS <- MSS[!phosphosignal == 0]
ck2_sites <- ck2_sites[!phosphosignal == 0]
phosphosignal <- phosphosignal[!phosphosignal == 0]

# Take logs and then plot
phosphosignal <- log(phosphosignal,base=10)

# Assign mutant labels to the phosphorylation signal and CK2 peptides
phosphosignal_mut <- phosphosignal
ck2_sites_mut <- ck2_sites

```

# Calculate motif score using a full PWM (20x13) derived from known substrates (Bachman et al., 2022 bioRxiv)

```{r}

# Fetch the CK2 PWM

lit_pwms <- readRDS('Bachmann_PWMs_disorder_weighted.rds')
ck2_pwm <- lit_pwms$`CSNK2A1, N=579`
ck2_pwm <- ck2_pwm[,2:14]

# scoring function for the motif score

mss_flank <- function(psite,pwm) {
  
  psite <- unlist(strsplit(psite,split=''))
  pwm <- as.matrix(pwm)
    
  #match.ic = apply(pwm, 2, function(col) sum(col * logb(20 * col), na.rm=T))
  
  c=0
  min=0
  max = 0
  
  # Ignore p0
  
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    
    # For missing amino acids, we score the '_' position
    # the same as we would for a minimum frequency AA
    
    if (aa == '_') {
      current <- min(pwm[,i])#*match.ic[i]
    } else {
      current <- pwm[rownames(pwm) %in% psite[i],i]#*match.ic[i]
    }
    
    minimum <- min(pwm[,i])#*match.ic[i]
    maximum <- max(pwm[,i])#*match.ic[i]
    
    c = c+current
    min = min+minimum
    max = max+maximum
    
  }
  
  mss = (c-min)/(max-min)
  
  #print(c)
  #print(min)
  #print(max)
  #print(match.ic)
  
  return(unname(mss))
  
}

# score each of the wt sites

mss_vec <- NULL

for (i in 1:length(ck2_sites_wt)) {
  
  mss <- mss_flank(ck2_sites_wt[i],ck2_pwm)
  mss_vec <- c(mss_vec,mss)
  
}

lit_ck2_wt_sites <- mss_vec

# score each of the mutant sites

mss_vec <- NULL

for (i in 1:length(ck2_sites_mut)) {
  
  mss <- mss_flank(ck2_sites_mut[i],ck2_pwm)
  mss_vec <- c(mss_vec,mss)
  
}

lit_ck2_mut_sites <- mss_vec

```

# Calculate motif score using a full PWM (20x13) derived from in vitro MS/MS data (Sugiyama et al., 2019 Sci. Reports)

```{r}

# Fetch the CK2 PWM

msms_pwms <- readRDS('Sugiyama_PWMs_disorder_weighted.rds')
ck2_pwm <- msms_pwms$`CK2a2, N=364`
ck2_pwm <- ck2_pwm[,2:14]

# scoring function for the motif score

mss_flank <- function(psite,pwm) {
  
  psite <- unlist(strsplit(psite,split=''))
  pwm <- as.matrix(pwm)
    
  #match.ic = apply(pwm, 2, function(col) sum(col * logb(20 * col), na.rm=T))
  
  c=0
  min=0
  max = 0
  
  # Ignore p0
  
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    
    # For missing amino acids, we score the '_' position
    # the same as we would for a minimum frequency AA
    
    if (aa == '_') {
      current <- min(pwm[,i])#*match.ic[i]
    } else {
      current <- pwm[rownames(pwm) %in% psite[i],i]#*match.ic[i]
    }
    
    minimum <- min(pwm[,i])#*match.ic[i]
    maximum <- max(pwm[,i])#*match.ic[i]
    
    c = c+current
    min = min+minimum
    max = max+maximum
    
  }
  
  mss = (c-min)/(max-min)
  
  #print(c)
  #print(min)
  #print(max)
  #print(match.ic)
  
  return(unname(mss))
  
}

# score each of the wt sites

mss_vec <- NULL

for (i in 1:length(ck2_sites_wt)) {
  
  mss <- mss_flank(ck2_sites_wt[i],ck2_pwm)
  mss_vec <- c(mss_vec,mss)
  
}

msms_ck2_wt_sites <- mss_vec

# score each of the mutant sites

mss_vec <- NULL

for (i in 1:length(ck2_sites_mut)) {
  
  mss <- mss_flank(ck2_sites_mut[i],ck2_pwm)
  mss_vec <- c(mss_vec,mss)
  
}

msms_ck2_mut_sites <- mss_vec

```

# Calculate motif score using a full PWM (20x9) derived from the synthetic peptide library data presented in (Johnson et al., 2023 Nature)

```{r}

# Retrieve CK2 data and remove pS/pT/pY values, which we cannot use here

cantley_pwms <- read_xlsx('Cantley_PWMs.xlsx', skip=0, col_names=TRUE, sheet=4)
cantley_pwms <- as.data.frame(cantley_pwms)
rownames(cantley_pwms) <- cantley_pwms[,1]
cantley_pwms <- cantley_pwms[,-1]
cantley_ck2 <- matrix(unlist(cantley_pwms[rownames(cantley_pwms) == 'CK2A1',]),nrow=23)
colnames(cantley_ck2) <- c(-5,-4,-3,-2,-1,1,2,3,4)
rownames(cantley_ck2) <- substr(colnames(cantley_pwms)[1:23],3,3)
cantley_ck2 <- cantley_ck2[1:20,]
ck2_pwm <- cantley_ck2

# These PWMs go from -5 to +4 only and so we have to format the peptide strings accordingly

ck2_sites_cantley_wt <- paste(substr(ck2_sites_wt,2,6),substr(ck2_sites_wt,8,11),sep='')
ck2_sites_cantley_mut <- paste(substr(ck2_sites_mut,2,6),substr(ck2_sites_mut,8,11),sep='')

# scoring function for the motif score (modified to account for the change in peptide length)

mss_flank <- function(psite,pwm) {
  
  psite <- unlist(strsplit(psite,split=''))
  pwm <- as.matrix(pwm)
    
  #match.ic = apply(pwm, 2, function(col) sum(col * logb(20 * col), na.rm=T))
  
  c=0
  min=0
  max = 0
  
  # Ignore p0
  
  for (i in c(1:9)) {
    
    aa <- psite[i]  
    
    # For missing amino acids, we score the '_' position
    # the same as we would for a minimum frequency AA
    
    if (aa == '_') {
      current <- min(pwm[,i])#*match.ic[i]
    } else {
      current <- pwm[rownames(pwm) %in% psite[i],i]#*match.ic[i]
    }
    
    minimum <- min(pwm[,i])#*match.ic[i]
    maximum <- max(pwm[,i])#*match.ic[i]
    
    c = c+current
    min = min+minimum
    max = max+maximum
    
  }
  
  mss = (c-min)/(max-min)
  
  #print(c)
  #print(min)
  #print(max)
  #print(match.ic)
  
  return(unname(mss))
  
}

# score each of the wt sites

mss_vec <- NULL

for (i in 1:length(ck2_sites_cantley_wt)) {
  
  mss <- mss_flank(ck2_sites_cantley_wt[i],ck2_pwm)
  mss_vec <- c(mss_vec,mss)
  
}

cantley_ck2_wt_sites <- mss_vec

# score each of the mutant sites

mss_vec <- NULL

for (i in 1:length(ck2_sites_cantley_mut)) {
  
  mss <- mss_flank(ck2_sites_cantley_mut[i],ck2_pwm)
  mss_vec <- c(mss_vec,mss)
  
}

cantley_ck2_mut_sites <- mss_vec

```

## ggplot the scatter plots for the WT peptides

```{r}

# Lit PWM and WT sites

mss_df <- data.frame(lit_ck2_wt_sites,phosphosignal_wt)
colnames(mss_df) <- c('MSS','phosphosignal')

p <- ggplot(mss_df, aes(x=MSS, y=phosphosignal)) + 
  geom_point()

p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15), panel.border = element_rect(color="black", size=1.2, linetype="solid"))+theme(legend.position="none")

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")))
p <- p+theme(axis.text=element_text(size=7),axis.title.x=element_text(size=8.5,face="plain"),axis.title.y=element_text(size=8.5,face="plain"),plot.title=element_text(size=10,face='bold'))
p <- p+theme(legend.position="none")
p <- p + xlab('') + ylab('log10 (WT peptides)') + ggtitle('literature PWM')
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(panel.border= element_blank())

cor.test(mss_df[,1],mss_df[,2],method='spearman')
p <- p+geom_richtext(x=0.75, y=3.5, label='r<sub>s</sub> = 0.45',size=3.5,fontface='plain',col='black',label.color = NA)

lit_wt_gg <- p
lit_wt_df <- mss_df

# MS/MS PWM and WT sites

mss_df <- data.frame(msms_ck2_wt_sites,phosphosignal_wt)
colnames(mss_df) <- c('MSS','phosphosignal')

p <- ggplot(mss_df, aes(x=MSS, y=phosphosignal)) + 
  geom_point()

p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15), panel.border = element_rect(color="black", size=1.2, linetype="solid"))+theme(legend.position="none")

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")))
p <- p+theme(axis.text=element_text(size=7),axis.title.x=element_text(size=8.5,face="plain"),axis.title.y=element_text(size=8.5,face="plain"),plot.title=element_text(size=10,face='bold'))
p <- p+theme(legend.position="none")
p <- p + xlab('') + ylab('') + ggtitle('MS/MS PWM')
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(panel.border= element_blank())

cor.test(mss_df[,1],mss_df[,2],method='spearman')
p <- p+geom_richtext(x=0.80, y=3.5, label='r<sub>s</sub> = 0.49',size=3.5,fontface='plain',col='black',label.color = NA)

msms_wt_gg <- p
msms_wt_df <- mss_df

# peptide library PWM and WT sites

mss_df <- data.frame(cantley_ck2_wt_sites,phosphosignal_wt)
colnames(mss_df) <- c('MSS','phosphosignal')

p <- ggplot(mss_df, aes(x=MSS, y=phosphosignal)) + 
  geom_point()

p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15), panel.border = element_rect(color="black", size=1.2, linetype="solid"))+theme(legend.position="none")

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")))
p <- p+theme(axis.text=element_text(size=7),axis.title.x=element_text(size=8.5,face="plain"),axis.title.y=element_text(size=8.5,face="plain"),plot.title=element_text(size=10,face='bold'))
p <- p+theme(legend.position="none")
p <- p + xlab('') + ylab('()') + ggtitle('peptide library PWM')
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(panel.border= element_blank())

cor.test(mss_df[,1],mss_df[,2],method='spearman')
p <- p+geom_richtext(x=0.80, y=3.5, label='r<sub>s</sub> = 0.49',size=3.5,fontface='plain',col='black',label.color = NA)

cantley_wt_gg <- p
cantley_wt_df <- mss_df

```

## ggplot the scatter plots for the WT+mutant peptides

```{r}

# Lit PWM and mutant sites

mss_df <- data.frame(lit_ck2_mut_sites,phosphosignal_mut)
colnames(mss_df) <- c('MSS','phosphosignal')

p <- ggplot(mss_df, aes(x=MSS, y=phosphosignal)) + 
  geom_point()

p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15), panel.border = element_rect(color="black", size=1.2, linetype="solid"))+theme(legend.position="none")

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")))
p <- p+theme(axis.text=element_text(size=7),axis.title.x=element_text(size=8.5,face="plain"),axis.title.y=element_text(size=8.5,face="plain"),plot.title=element_text(size=10,face='bold'))
p <- p+theme(legend.position="none")
p <- p + xlab('') + ylab('log10 (WT+mut peptides)') + ggtitle('')
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(panel.border= element_blank())

cor.test(mss_df[,1],mss_df[,2],method='spearman')
p <- p+geom_richtext(x=0.8, y=3.5, label='r<sub>s</sub> = 0.56',size=3.5,fontface='plain',col='black',label.color = NA)

lit_mut_gg <- p
lit_mut_df <- mss_df

# MS/MS PWM and WT sites

mss_df <- data.frame(msms_ck2_mut_sites,phosphosignal_mut)
colnames(mss_df) <- c('MSS','phosphosignal')

p <- ggplot(mss_df, aes(x=MSS, y=phosphosignal)) + 
  geom_point()

p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15), panel.border = element_rect(color="black", size=1.2, linetype="solid"))+theme(legend.position="none")

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")))
p <- p+theme(axis.text=element_text(size=7),axis.title.x=element_text(size=8.5,face="plain"),axis.title.y=element_text(size=8.5,face="plain"),plot.title=element_text(size=10,face='bold'))
p <- p+theme(legend.position="none")
p <- p + xlab('Substrate quality') + ylab('') + ggtitle('')
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(panel.border= element_blank())

cor.test(mss_df[,1],mss_df[,2],method='spearman')
p <- p+geom_richtext(x=0.85, y=3.5, label='r<sub>s</sub> = 0.57',size=3.5,fontface='plain',col='black',label.color = NA)

msms_mut_gg <- p
msms_mut_df <- mss_df

# peptide library PWM and WT sites

mss_df <- data.frame(cantley_ck2_mut_sites,phosphosignal_mut)
colnames(mss_df) <- c('MSS','phosphosignal')

p <- ggplot(mss_df, aes(x=MSS, y=phosphosignal)) + 
  geom_point()

p <- p+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15), panel.border = element_rect(color="black", size=1.2, linetype="solid"))+theme(legend.position="none")

# Ticks
p <- p+theme(axis.ticks.length=unit(-0.10, "cm"), axis.text.x = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")), axis.text.y = element_text(margin=unit(c(0.2,0.2,0.2,0.2), "cm")))
p <- p+theme(axis.text=element_text(size=7),axis.title.x=element_text(size=8.5,face="plain"),axis.title.y=element_text(size=8.5,face="plain"),plot.title=element_text(size=10,face='bold'))
p <- p+theme(legend.position="none")
p <- p + xlab('') + ylab('') + ggtitle('')
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(panel.border= element_blank())

cor.test(mss_df[,1],mss_df[,2],method='spearman')
p <- p+geom_richtext(x=0.80, y=3.5, label='r<sub>s</sub> = 0.48',size=3.5,fontface='plain',col='black',label.color = NA)

cantley_mut_gg <- p
cantley_mut_df <- mss_df


```

# Cowplot (put it all together)

```{r}

library(cowplot)

sup_fig_3 <- plot_grid(lit_wt_gg,msms_wt_gg,cantley_wt_gg,lit_mut_gg,msms_mut_gg,cantley_mut_gg,ncol=3,rel_widths=c(1,1,1),labels=c('a','b','c','','',''))
ggsave(file='Supplementary_figure_3.pdf', plot=sup_fig_3, width=10, height=5.8)

```

# Compile supplementary table 2 from the constituent data frames

```{r}

library(xlsx)

df <- data.frame(ck2_sites_wt,fig_1d_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'WT_peptides_DE_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

####

df <- data.frame(ck2_sites_wt,lit_wt_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'WT_peptides_lit_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

###

df <- data.frame(ck2_sites_wt,msms_wt_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'WT_peptides_MSMS_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

###

df <- data.frame(ck2_sites_wt,cantley_wt_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'WT_peptides_peplibrary_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

###

df <- data.frame(ck2_sites_mut, sup_fig2_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'mut_peptides_DE_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

###

df <- data.frame(ck2_sites_mut, lit_mut_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'mut_peptides_lit_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

###

df <- data.frame(ck2_sites_mut, msms_mut_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'mut_peptides_msms_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

###

df <- data.frame(ck2_sites_mut, cantley_mut_df)
colnames(df) <- c('peptide','MSS','log10(phosphorylation)')

write.xlsx(df, 'Supplementary_table_2.xlsx', sheetName = 'mut_peptides_peplibrary_PWM', 
  col.names = TRUE, row.names = FALSE, append = TRUE)

```


