---
title: "Figure_2e"
author: "David Bradley"
date: "10/12/2020"
output: html_document
---

# GGplot font

```{r}

# https://stackoverflow.com/questions/27689222/changing-fonts-for-graphs-in-r

library(extrafont)

font_import()
loadfonts()       #Register fonts for Windows bitmap output
fonts() 

```

# Retrieve the CK2 substrates and model specificity

```{r}

library(readr)
library(seqinr)
library(ggseqlogo)
library(gdata)
library(stringr)

# Data comes from Bachmann et al., 2019 (https://www.biorxiv.org/content/10.1101/822668v3)

prot_mapper <- read_csv('export.csv')

# Kinases only

prot_mapper <- prot_mapper[unlist(prot_mapper[,5]) == TRUE,]

# Retrieve upstream kinase, substrate, and substrate position

ksr <- prot_mapper[,c(3,4,6,9,10)]

# We exclude KSRs that are supported only by the text-mining tools

curated <- c('psp','signor','pid','bel','reactome')

source_list <- strsplit(unlist(ksr[,5]),split=',')

number_curated <- lapply(source_list, function(x) length(intersect(x,curated)))

ksr <- ksr[which(number_curated > 0),]

# Correct formatting problems

for (i in 1:nrow(ksr)) {
 
  if(is.na(ksr[i,2])) {ksr[i,2] <- ksr[i,1]} 
  
}

# Retrieve CSNK2A1 and CSNK2A2 substrates

csnk2a1_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A1') == TRUE),]
csnk2a2_ksr <- ksr[which(unlist(ksr[,2] == 'CSNK2A2') == TRUE),]
csnk2a_ksr <- unique(rbind(csnk2a1_ksr,csnk2a2_ksr))

# Human proteome

hs_rp <- read.fasta('hs_proteome_uniprot.fasta', seqtype='AA')
hs_ids <- rapply(strsplit(names(hs_rp),split='\\|'),function(x) x[2])

# Retrieve the sequences

psite_vec <- NULL
index_vec <- NULL

for (i in 1:nrow(csnk2a_ksr)) {
  
  print(i)
  
  # Retrieve the first element only
  uniprot_id <- unname(unlist(csnk2a_ksr[i,3]))
  site_pos <- unname(unlist(csnk2a_ksr[i,4]))
  
  # Loop for ambiguous psite positions
  
  for (j in 1:length(uniprot_id)) {
    
    pos <- as.numeric(site_pos)
    
    # Substrate sequence
  
    sub_seq <- unlist(getSequence(hs_rp[hs_ids %in% uniprot_id[j]]))
    
    if (length(sub_seq) == 0) {next}
    
    # Phosphosite position
    
    if (pos < 8) {
      
      gap <- paste(rep('_',(8-pos)),collapse='')
      psite <- paste(sub_seq[1:(pos+7)],collapse='')
      psite <- paste(gap,psite,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else if ((pos+7) > length(sub_seq)) {
      
      psite <- paste(sub_seq[(pos-7):(length(sub_seq))],collapse='')
      gap <- paste(rep('_',(pos+7)-length(sub_seq)),collapse='')
      psite <- paste(psite,gap,sep='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    } else {
      
      psite <- paste(sub_seq[(pos-7):(pos+7)],collapse='')
      if (!substr(psite,8,8) %in% c('S','T','Y')) {next}
      psite_vec <- c(psite_vec, psite)
      index_vec <- c(index_vec, i)
      
    }
  } 
}

# Remove potential heirarchical sites
bachmann_vec <- psite_vec
p1s <- rapply(strsplit(psite_vec,split=''), function(x) x[9]) == 'S'
p3s <- rapply(strsplit(psite_vec,split=''), function(x) x[11]) == 'S'
psite_no_heira <- psite_vec[!(p1s|p3s)]

# Retrieve the corresponding database records

csnk2a_ksr_no_heira <- csnk2a_ksr[index_vec[!(p1s|p3s)],]
csnk2a_ksr_no_heira <- cbind(csnk2a_ksr_no_heira,psite_no_heira)

# Generate the PWM

source('match-tm.r')

csnk2_pwm <- makePWM(substr(psite_no_heira,2,14))

# Scoring function for the calculation of CK2 substrate quality

ED_weight <- function(psite,pwm) {
  
  neg_row <- apply(pwm[rownames(pwm) %in% c('D','E'),],2,sum)
  pwm <- pwm[!rownames(pwm) %in% c('D','E'),]
  pwm <- rbind(pwm,neg_row)
  rownames(pwm)[nrow(pwm)] <- 'DE'
  
  psite <- unlist(strsplit(psite,split=''))
  
  c=0
  min=0
  
  # Ignore the central residue
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    if (aa %in% c('D','E')) {aa <- 'DE'}
    # Skip if the amino acid is not D or E
    if (!(aa %in% c('DE'))) {next}
    
    current <- pwm[rownames(pwm) %in% aa,i]
    c = c+current
  
  }
  
  maximum <- sum(apply(pwm[,c(1:6,8:13)],2,max))
  mss = c/maximum
  
  return(unname(mss))
  
}

```

# Analyse the human cell cycle data (Olsen et al., 2010) using known CK2 substrates

```{r}

### Correct for duplicated records

csnk2a_ksr_no_heira <- csnk2a_ksr_no_heira[!duplicated(csnk2a_ksr_no_heira[c(3,4)]),]

library(stringr)
library(readxl)
library(seqinr)
library(gdata)

source('match-tm.r')

# Read in the file for the human cell cycle data

Olsen_S2 <- read.xls('Olsen_2010_S2.xls',sheet=1, stringsAsFactors=FALSE)

# Class I sites only

Olsen_S2 <- Olsen_S2[grep('YES',Olsen_S2[,5]),]

# Single phosphosite only

Olsen_S2 <- Olsen_S2[Olsen_S2[,16] == 1,]

# Retrieve sequence window

Olsen_sites <- Olsen_S2[,18]

# Known CSNK2 substrates
csnk2_sub <- unname(unlist(csnk2a_ksr_no_heira[,3]))

# Retrieve the Olsen UniProt accessions
Olsen_S2_UP <- strsplit(Olsen_S2[,2],split=';')

col1 <- unlist(Olsen_S2_UP)[unlist(lapply(Olsen_S2_UP, function(x) x %in% csnk2_sub))]

i_vec <- NULL

for (i in 1:length(Olsen_S2_UP)) {
  
 i_vec <- c(i_vec, rep(i,length(Olsen_S2_UP[[i]])))
  
}

pos <- as.numeric(Olsen_S2[i_vec,3])

col2 <- pos[unlist(lapply(Olsen_S2_UP, function(x) x %in% csnk2_sub))]

col1_col2 <- cbind(col1,col2)

match_dex <- match(paste(unlist(csnk2a_ksr_no_heira[,3]),unlist(csnk2a_ksr_no_heira[,4])),paste(col1,col2))

olsen_ck2_sub <- csnk2a_ksr_no_heira[!is.na(match_dex),]

# Match the known CK2 substrates to the sites recorded in the cell cycle phosphoproteome data

grep_vec <- NULL

for (i in 1:nrow(olsen_ck2_sub)) {
  
  sub <- unlist(unname(olsen_ck2_sub[i,3]))
  pos <- unlist(unname(olsen_ck2_sub[i,4]))
  
  grep1 <- grep(sub,Olsen_S2[,2])
  grep2 <- grep(paste0("^",pos,"$"),Olsen_S2[,3])
  
  grep <- intersect(grep1,grep2)
  
  if (length(grep) > 1) {stop()}
    
  grep_vec <- c(grep_vec, grep)
  
}

# Generate the CK2 sites

Olsen_S2_ck2 <- Olsen_S2[grep_vec,]

Olsen_S2_ck2_ratios <- Olsen_S2_ck2[,6:11]

ck2_sites <- Olsen_S2_ck2[,18]

# Scoring function for calculating the CK2 substrate quality

ED_weight <- function(psite,pwm) {
  
  neg_row <- apply(pwm[rownames(pwm) %in% c('D','E'),],2,sum)
  pwm <- pwm[!rownames(pwm) %in% c('D','E'),]
  pwm <- rbind(pwm,neg_row)
  rownames(pwm)[nrow(pwm)] <- 'DE'
  
  psite <- unlist(strsplit(psite,split=''))
  
  c=0
  min=0
  
  # Ignore the central residue
  for (i in c(1:6,8:13)) {
    
    aa <- psite[i]  
    if (aa %in% c('D','E')) {aa <- 'DE'}
    # Skip if the amino acid is not D or E
    if (!(aa %in% c('DE'))) {next}
    
    current <- pwm[rownames(pwm) %in% aa,i]
    c = c+current
  
  }
  
  maximum <- sum(apply(pwm[,c(1:6,8:13)],2,max))
  mss = c/maximum
  
  return(unname(mss))
  
}

csnk2_pwm <- makePWM(substr(psite_no_heira,2,14))

mss_vec <- NULL

for (i in 1:length(ck2_sites)) {
  
  mss <- NULL
  ck2_site <- ck2_sites[i]
  mss <- ED_weight(ck2_site,csnk2_pwm)
  if (length(mss) == 0) {break}
  mss_vec <- c(mss_vec, mss)
  
}

Olsen_S2_ck2_ratios <- cbind(ck2_sites, Olsen_S2_ck2_ratios)

# Filter out phoshosites with missing values for a cell cycle stage

library(tidyr)

ck2_time_course <- Olsen_S2_ck2_ratios
na_count <- apply(Olsen_S2_ck2_ratios , 1, function(x) length(which(is.na(x))))
ck2_time_course  <- ck2_time_course[na_count < 1,]
ck2_time_course <- cbind(rep('Time',nrow(ck2_time_course)),ck2_time_course)

# Mark out the two examples given in the MS

ck2_time_course[ck2_time_course[,2] == 'PKRKVSSAEGAAK',1] <- 'Minimal'
ck2_time_course[ck2_time_course[,2] == 'WDHIEVSDDEDET',1] <- 'Optimal'

colnames(ck2_time_course) <- rapply(strsplit(colnames(ck2_time_course),split='Log2.'), function(x) x[length(x)])

# Put mitosis last

ck2_time_course <- ck2_time_course[,c(1,2,4,5,6,7,8,3)]

# Plot the time profile of CK2 substrates along the different stages of the cell cycle while highlighting two interesting examples
# that are described in the paper

ggplot_long <- pivot_longer(ck2_time_course, cols = c(colnames(ck2_time_course)[3:ncol(ck2_time_course)]))
colnames(ggplot_long) <- c('Motif','Sequence','Time','Ratio')
ggplot_long$Time <- factor(ggplot_long$Time, levels = unique(ggplot_long$Time))
ggplot_long$Motif <- factor(ggplot_long$Motif, levels = c('Optimal','Time','Minimal'))

ratio <- as.numeric(unlist(ggplot_long[,4]))
ggplot_long <- ggplot_long[!is.na(ratio),]

ggplot_long[,4] <- as.numeric(unlist(ggplot_long[,4]))
g <- ggplot(data=ggplot_long, aes(x=Time, y=Ratio, group=Sequence)) +
  #geom_line(color = 'lightblue', lwd=1) +
  geom_line(aes(color = Motif), lwd=1) +
  geom_point(aes(color = Motif, alpha = Motif),size=2.5) +
  ylim(-4,6)
g <- g+theme_bw() + theme(text=element_text(family="Ubuntu Light", face="plain", size=15),legend.position = "none")
g <- g + scale_color_manual(values=c(colors()[132],colors()[246], colors()[430]))
g <- g + scale_alpha_manual(values=c(1,0,1))
g <- g + theme(axis.text=element_text(size=9),axis.title=element_text(size=11,face="plain"),plot.title=element_text(size=13,face='bold'))
g <- g + ylab("SILAC ratio") + xlab("") + ggtitle('')
g <- g+theme(plot.title = element_text(hjust = 0.5))
g

g <- g+geom_text(data=data.frame(), aes(x=c(1.55,1.45), y=c(3.8,-1.0), label=c('HMGN1 S8 \n (0.09)','Cdc37 S13 \n (0.68)')), col=c(colors()[430],colors()[132]), size=3.0, fontface='bold', inherit.aes = FALSE)
g

figure_2e <- g
mss_data_figure_2e <- ggplot_long

ggsave(file='Figure_2e.pdf', plot=g, width=6.0, height=4.8)

```
```